lib/src/web_socket/browser/sockjs.dart

```diff
@@ -61,46 +60,76 @@ class SockJSSocket extends CommonWSocket implements WSocket {
-      _socket.send(message);
-    }, onError: handleOutgoingError, onDone: handleOutgoingDone);
-
-    // Map events from the underlying socket to the incoming controller.
-    incoming = new StreamController();
-    _socket.onMessage.listen((messageEvent) {
-      // Pipe messages from the socket through to the stream.
-      incoming.add(messageEvent.data);
+  sockjs.Client _webSocket;
+
+  SockJSSocket._(this._webSocket, Future webSocketClosed) : super() {
+    webSocketClosed.then((closeEvent) {
+      closeCode = closeEvent.code;
+      closeReason = closeEvent.reason;
+      onIncomingDone();
     });
-    _socketClosed.then((closeEvent) {
-      // Now that the socket has closed, capture the close code and reason.
-      var wCloseEvent =
-          new WSocketCloseEvent(closeEvent.code, closeEvent.reason);
-      handleSocketDone(wCloseEvent);
+
+    // Note: We don't listen to the SockJS client for messages immediately like
+    // we do with the native WebSockets. This is because the event streams from
+    // the SockJS client are all drawn from a single broadcast stream. To make
+    // it act like a single subscription stream (and thus make it fit the
+    // interface of a standard Stream), we create a subscription when a consumer
+    // listens to this WSocket instance, cancel that subscription when the
+    // consumer's subscription is paused, and re-listen when the consumer
+    // resumes listening. See [onIncomingListen], [onIncomingPause], and
+    // [onIncomingResume].
+
+    // Additional note: the SockJS Client has no error stream, so no need to
+    // listen to for errors.
+  }
+
+  @override
+  void closeWebSocket(int code, String reason) {
+    _webSocket.close(code, reason);
+  }
+
+  @override
+  void onIncomingError(error, [StackTrace stackTrace]) {
+    shutDown(error: error, stackTrace: stackTrace);
+  }
+
+  @override
+  void onIncomingListen() {
+    // When this [WSocket] instance is listened to, start listening to the
+    // SockJS client's broadcast stream.
+    webSocketSubscription = _webSocket.onMessage.listen((messageEvent) {
+      onIncomingData(messageEvent.data);
+    });
+  }
+
+  @override
+  void onIncomingPause() {
+    // When this [WSocket]'s subscription is paused, cancel the subscription to
+    // the SockJS client's broadcast stream. This is the recommended behavior
+    // when proxying a subscription to a broadcast stream. This effectively
+    // prevents buffering events indefinitely (a possible memory leak) by
+    // canceling the subscription altogether. When the subscription to this
+    // [WSocket] instance is resumed, we will re-subscribe.
+    webSocketSubscription.cancel();
+  }
+
+  @override
+  void onIncomingResume() {
+    // Resubscribe to the SockJS client's broadcast stream to effectively resume
+    // the consumer's subscription to this [WSocket] instance.
+    webSocketSubscription = _webSocket.onMessage.listen((messageEvent) {
```

Comment:
is there any risk of orphaning an existing subscription since this doesn't check / cancel any existing subscription before re-assigning?  or is this only used internally and we're confident that it's always used properly?